lib/stdlib/src/supervisor.erl

```diff
@@ -1516,20 +1516,44 @@ shutdown(#child{pid=Pid, shutdown=Time} = Child) ->
                     normal when not (?is_permanent(Child)) ->
                         ok;
                     Reason ->
+                        %% killed is expected to become an error as
+                        %% this function clause tries to terminate
+                        %% process not brutally
                         {error, Reason}
                 end
         end
     end.
 
-unlink_flush(Pid, DefaultReason) ->
+unlink_flush(Pid, Reason) ->
+    {links, Links} = process_info(self(), links),
+    case lists:member(Pid, Links) of
+        true ->
+            do_unlink_flush(Pid, Reason);
+        false  -> %% Naughty child
+            do_flush(Pid, Reason)
+    end.
```

Comment:
A question re this, though. Is it guaranteed that, if a process was linked to the supervisor and `process_info(self(), links)` does not return its pid any more, the respective `'EXIT'` message is then in the mailbox of the supervisor? Because if not, there still remains an (admittedly very tiny) window for a race condition.

Reply:
> A question re this, though. Is it guaranteed that, if a process was linked to the supervisor and `process_info(self(), links)` does not return its pid any more, the respective `'EXIT'` message is then in the mailbox of the supervisor?

Assuming that the link has been there since the child was spawned, yes.